eval.go

```diff
@@ -587,7 +577,29 @@ func (d *indexData) gatherBranches(docID uint32, mt matchTree, known map[matchTr
 			id <<= 1
 			mask >>= 1
 		}
+
+		return branches
 	}
+
+	foundBranchQuery := false
+	var branches []string
+	visitMatches(mt, known, func(mt matchTree) {
+		bq, ok := mt.(*branchQueryMatchTree)
+		if !ok {
+			return
+		}
+		foundBranchQuery = true
+		mask := bq.masks[repoIdx] & bq.fileMasks[docID]
```

Comment:
It took jumping around the code a bit for me to validate this is correct. What is weird in the code is that bq.fileMasks is actually referencing indexData.fileBranchMasks. Maybe this could would be easier to understand if you added a function to branchQueryMatchTree which returned the current mask? IE docID is already part of the branch match tree state, so you could just having something with the signature `branchQueryMatchTree.branchMask() uint32`. That branchMask could then also be used by `branchQueryMatchTree.matches`.

Reply:
Funny enough the test failed once I made the changes because it ignored the state of the branchQueryMatchTree. I updated the test to do an actual search and now things work as expected.